### PR TITLE
Update Trinity to v2.1.1 and add post-assembly scripts

### PR DESCRIPTION
**Pull request by abretaud**
Hi,
Here's an updated wrapper for Trinity, and some wrappers for scripts bundled in the trinity package.
I had some chunks of these new wrappers waiting on my hard drive, I thought it could be useful to update them and upload them to IUC.
With this, it should now be possible to do most of the [Trinity tutorial](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Post-Transcriptome-Assembly-Downstream-Analyses)
Anthony

### CODE REVIEW

**Comment by bgruening** [comment]
@abretaud can you please check the travis error, there is some invalid XML in this PR.
Also please make sure this is the correct PR. It seems here is many more included than trinity, which is great, just checking!

**Comment by abretaud** [comment]
Oops, I had just accidently deleted an xml tag, it should be ok now
Yes, it's the right PR, there is more than just trinity because rsem is no longer distributed in the trinity package, and kallisto is used by one of the new wrapper

File: packages/package_trinity_2_0_6/tool_dependencies.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0"?>
 <tool_dependency>
-    <package name="ncurses" version="5.9">
-        <repository name="package_ncurses_5_9" owner="iuc" prior_installation_required="True" />
+    <package name="ncurses" version="6.0">
```
Comment (bgruening) [inline packages/package_trinity_2_0_6/tool_dependencies.xml]: is this change really needed?

File: packages/package_trinity_2_1_1/tool_dependencies.xml [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                        https://depot.galaxyproject.org/software/reshape2/reshape2_1.4.1_src_all.tar.gz
+                    </package>
+                    <package sha256sum="36f8bec5831ceedde3d3ab21d9db87cf1a32133a04f46e1803c833a510b8d17d">
+                        http://bioconductor.org/packages/release/bioc/src/contrib/qvalue_2.2.2.tar.gz
+                    </package>
+                    <package sha256sum="b4752329457c250200171603b355bf9b4b64c8767777a233de7e7fb2d33fe3b7">
+                        http://cran.rstudio.com/src/contrib/amap_0.8-14.tar.gz
+                    </package>
+                    <package sha256sum="0f67391d3fd6bc970293b5eb11dcd14f15686b61bce1c6da03e16dc3a60e93f2">
+                        http://www.bioconductor.org/packages/release/bioc/src/contrib/ctc_1.44.0.tar.gz
```
Comment (bgruening) [inline packages/package_trinity_2_1_1/tool_dependencies.xml]: Can we get those into cargo_port?

File: tools/trinity/run_DE_analysis.xml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    </stdio>
+    <command><![CDATA[
+        
+        ln -s ${matrix} input.matrix
+        
+        &&
+        
+        run_DE_analysis.pl
+        
+        --matrix input.matrix
```
Comment (bgruening) [inline tools/trinity/run_DE_analysis.xml]: quote

File: tools/trinity/run_DE_analysis.xml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        --samples_file ${samples}
```
Comment (bgruening) [inline tools/trinity/run_DE_analysis.xml]: quote

File: tools/trinity/run_DE_analysis.xml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        
+        --output results
+        
+    ]]></command>
+    <inputs>
+        <param format="tabular" name="matrix" type="data" label="Expression matrix" help="output of abundance_estimates_to_matrix tool"/>
+        <param format="tabular" name="samples" type="data" label="Sample description" help="file describing samples and replicates"/>
+        
+        <conditional name="method_choice">
+            <param type="select" name="method" label="Differential analysis method">
```
Comment (bgruening) [inline tools/trinity/run_DE_analysis.xml]: edgeR, voom are not in the dependency list ....

File: tools/trinity/samples_qccheck.xml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        <exit_code range="1:"/>
+    </stdio>
+    <command><![CDATA[
+        
+        ln -s ${matrix} input.matrix
+        
+        &&
+        
+        PtR
+        --matrix input.matrix
```
Comment (bgruening) [inline tools/trinity/samples_qccheck.xml]: quote

File: tools/trinity/samples_qccheck.xml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        
+        --samples ${samples}
```
Comment (bgruening) [inline tools/trinity/samples_qccheck.xml]: quote

File: tools/trinity/samples_qccheck.xml [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        
+        --CPM --log2 --compare_replicates
+        
+        &&
+        
+        PtR
+        --matrix input.matrix
+        
+        --samples ${samples}
```
Comment (bgruening) [inline tools/trinity/samples_qccheck.xml]: quote

File: tools/trinity/tool_dependencies.xml [hunk 9]
```diff
@@ -1,14 +1,35 @@
...
+        <repository name="package_rsem_1_2_28" owner="iuc" />
+    </package>
+    <package name="kallisto" version="0.42.4">
+        <repository name="package_kallisto_0_42_4" owner="iuc" />
+    </package>
+    <package name="express" version="1.5.1">
+        <repository name="package_express_1_5_1" owner="iuc" />
+    </package>
+    <package name="limma" version="3.25.3">
+        <repository name="package_r_limma_3_25_3" owner="iuc" />
```
Comment (bgruening) [inline tools/trinity/tool_dependencies.xml]: is voom part of limma?

**Comment by nsoranzo** [comment]
I'm not sure I like the addition of kallisto as a dependency, since it is notoriously non-free software: https://pachterlab.github.io/kallisto/license.html

**Comment by abretaud** [comment]
Thanks for reviewing!

- Voom is indeed part of the limma package.
- I added the missing packages to cargo_port: [PR49](https://github.com/galaxyproject/cargo-port/pull/49)
- For the dependency on ncurses 6.0 for trinity 2.0.6: this was a temporary fix for some compilation problems I had with ncurses 5.9. I have switched back to ncurse 5.9 which I fixed instead.
- I wasn't aware of the licence problem with kallisto... It is one of the methods that can be used by trinity, so it's not a _strict_ requirement. Is there a policy or an alternative for this kind of dependency?

**Comment by nsoranzo** [comment]
> I wasn't aware of the licence problem with kallisto... It is one of the methods that can be used by trinity, so it's not a strict requirement. Is there a policy or an alternative for this kind of dependency?

@abretaud There is no official policy, but I personally don't like the addition of a non-free and otherwise optional requirement to a FOSS tool wrapper.
On a side note, [Sailfish](https://www.cs.cmu.edu/~ckingsf/software/sailfish/) is a valid alternative to kallisto, but unfortunately it is not supported by Trinity yet.

**Comment by bgruening** [comment]
I'm also +/-0 for including kallisto. I personally don't support such tools, but we do not have any official policy here. @erasche @peterjc @mvdbeek @yhoogstrate what do you think?

**Comment by mvdbeek** [comment]
Not sure on that issue, I agree a more liberal license would be better.
```
Permission to use, copy, modify, and distribute this software and its documentation for educational and research not-for-profit purposes, without fee and without a signed licensing agreement, is hereby granted, provided that the above copyright notice, this paragraph and the following two paragraphs appear in all copies, modifications, and distributions.
```

So if that license is prominently included, I'd be + 0.5 for including kalisto.
Add another 0.5 is kalisto is really needed for trinity.

**Comment by peterjc** [comment]
Any Kallisto tool-shed package would be available to anyone (including for profit commercial users), so I think we'd have to ask for an special permission from the authors in order to package Kallisto for the Tool Shed.

**Comment by erasche** [comment]
@bgruening we've hit this in the past so we need an official policy.

I'd argue that we're fine to distribute (or at least "provide a script which installs" thereby letting *them* distribute). But definitely something we shouldn't put in the cargo port.

As for commercial users ... are we ok by just making the notice really large/loud and pointing to their download page regarding licensing?

I think this is a common enough problem that the TS should indicate the license and make it obvious where there will be issues.

**Comment by peterjc** [comment]
We do not have permission to distribute or redistribute (cache) Kallisto. It might be OK as long as we are just providing a script/recipe to install it which downloads directly from the authors' server? I think it would be safer to ask rather than just doing this.

I absolutely agree visibility of 3rd party licences is something the Tool Shed could be extended to deal with (beyond the README file alone). I've wrapped half a dozen non-free tools and always opted to require the SysAdmin manually install the dependencies, even when an automated installation would be possible for certain classes of user (like here with Kallisto).

**Comment by abretaud** [comment]
I just sent an email to the authors on the kallisto-sleuth mailing list

**Comment by erasche** [comment]
@peterjc yes, sorry, that's what I intended. Not redistributing for them.

File: packages/package_trinity_2_0_6/tool_dependencies.xml [hunk 10]
```diff
@@ -13,8 +13,8 @@
                     https://github.com/trinityrnaseq/trinityrnaseq/archive/v2.0.6.tar.gz
                 </action>
                 <action type="set_environment_for_install">
-                    <repository name="package_ncurses_5_9"  owner="iuc" prior_installation_required="True">
-                        <package name="ncurses" version="5.9" />
+                    <repository name="package_ncurses_6_0"  owner="iuc" prior_installation_required="True">
+                        <package name="ncurses" version="6.0" />
```
Comment (nsoranzo) [inline packages/package_trinity_2_0_6/tool_dependencies.xml]: Can you revert also this changes?

**Comment by abretaud** [comment]
As I don't have yet an answer from the authors, maybe I can move the kallisto dependency (and corresponding options in trinity) into another PR, and move on for the rest of this PR?

**Comment by nsoranzo** [comment]
@abretaud I was going to suggest the same.

**Comment by abretaud** [comment]
That's it, the kallisto problem is now in #613

**Comment by nsoranzo** [comment]
:+1: Thanks @abretaud!

**Comment by bgruening** [comment]
@abretaud this is great.
I have a rather unusal request and its ok if this you say no because you want to finally merge it.
All of your dependencies besides `edger` are already in conda ... so if we remove this line https://github.com/galaxyproject/tools-iuc/blob/master/.tt_blacklist#L40 we could test this PR with planemo test in travis.

edger is called `bioconductor-edger` in conda and available in version 3.12.0. So to fix this either we could create a 3.12.0 edger tool_dependencies.xml file or try to backport 3.12.0 to conda. I would prefer the first one.

What do you think?

**Comment by abretaud** [comment]
Will it work with like this? The conda stuff looks really cool!

File: tools/trinity/tool_dependencies.xml [hunk 11]
```diff
@@ -1,14 +1,35 @@
...
     </package>
+    <package name="bioconductor-edger" version="3.12.0">
+        <repository name="package_r_edger_3_12_0" owner="iuc" />
+    </package>
+    <package name="rsem" version="1.2.28">
+        <repository name="package_rsem_1_2_28" owner="iuc" />
+    </package>
+    <package name="kallisto" version="0.42.4">
+        <repository name="package_kallisto_0_42_4" owner="iuc" />
+    </package>
```
Comment (nsoranzo) [inline tools/trinity/tool_dependencies.xml]: You added kallisto again :laughing:

**Comment by abretaud** [comment]
Oops, sorry! Sometimes writing code in a rush is not a good idea :/

File: tools/trinity/abundance_estimates_to_matrix.xml [hunk 12]
```diff
@@ -0,0 +1,205 @@
+<tool id="abundance_estimates_to_matrix" name="Build expression matrix" version="2.1.1">
+    <description>for a de novo assembly of RNA-Seq data by Trinity</description>
+    <requirements>
+        <requirement type="package" version="2.1.1">trinity</requirement>
+        <requirement type="package" version="1.1.2">bowtie</requirement>
+        <requirement type="package" version="1.2">samtools</requirement>
+        <requirement type="package" version="3.11.0">edger</requirement>
```
Comment (nsoranzo) [inline tools/trinity/abundance_estimates_to_matrix.xml]: 3.12.0

File: tools/trinity/run_DE_analysis.xml [hunk 13]
```diff
@@ -0,0 +1,143 @@
+<tool id="run_DE_analysis" name="Differential expression analysis" version="2.1.1">
+    <description>using a Trinity assembly</description>
+    <requirements>
+        <requirement type="package" version="2.1.1">trinity</requirement>
+        <!-- odering is crucial, otherwise R will override the ENV variables from deseq2 -->
+        <requirement type="package" version="1.8.2">deseq2</requirement>
+        <requirement type="package" version="3.11.0">edger</requirement>
```
Comment (nsoranzo) [inline tools/trinity/run_DE_analysis.xml]: 3.12.0

File: packages/package_r_edger_3_12_0/tool_dependencies.xml [hunk 14]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
+<tool_dependency>
+    <package name="R" version="3.1.2">
+        <repository name="package_r_3_1_2" owner="iuc" prior_installation_required="True" />
+    </package>
+    <package name="bioconductor-edger" version="3.12.0">
```
Comment (nsoranzo) [inline packages/package_r_edger_3_12_0/tool_dependencies.xml]: I suppose the name here can remain `edger`, but @bgruening knows better.

**Comment by abretaud** [comment]
I'll change to edger instead of bioconductor-edger if @bgruening confirms that it's ok

File: packages/package_r_edger_3_12_0/tool_dependencies.xml (hunk 14 above)
Comment (bgruening) [inline packages/package_r_edger_3_12_0/tool_dependencies.xml]: Actually it needs to be bioconductor-edger to fit the bioconda package. For the transition time I think it's ok to adopt our packages also to the bioconductor naming scheme.

**Comment by bgruening** [comment]
@abretaud this is awesome. Can you remove this line https://github.com/galaxyproject/tools-iuc/blob/master/.tt_blacklist#L40 and we will see if testing works here :)

File: packages/package_r_edger_3_12_0/tool_dependencies.xml (hunk 14 above)
Comment (nsoranzo) [inline packages/package_r_edger_3_12_0/tool_dependencies.xml]: @bgruening I thought of something like https://github.com/galaxyproject/tools-iuc/pull/574/files#diff-faa05c6bd08139662c8a4b6ed27b8ea4

Comment (bgruening) [inline packages/package_r_edger_3_12_0/tool_dependencies.xml]: Possible, than the first one fails and Galaxy will pick up the second one. But if we have the chance to change the naming ... do you think it's to confusing because the package name does not match the dependency name?

Comment (nsoranzo) [inline packages/package_r_edger_3_12_0/tool_dependencies.xml]: Yes, and it's different from the previous version.

Comment (bgruening) [inline packages/package_r_edger_3_12_0/tool_dependencies.xml]: I don't mind. We can also include 2 requirements (one conda fallback). At some point we need to adopt the bioconda naming schema.

**Comment by abretaud** [comment]
I'm trying to fix the trinity bioconda recipe to include the missing scripts in the PATH.
The simplest way would be to symlink those scripts in the $PREFIX/bin dir, _but_ it doesn't work because these scripts are expecting to find some other files in subdirectories.
Do you know what is the best option to fix this? (patching the trinity code (argh), somehow customizing the path?)

For example $TRINITY_HOME/Analysis/DifferentialExpression/run_DE_analysis.pl will fail if launched using a symlink

**Comment by bgruening** [comment]
@abretaud does exporting PATH to this folder help in build.sh?

**Comment by nsoranzo** [comment]
@abretaud Can you paste the error here or on a gist? I'd say patch the Perl script and push the fix upstream on https://github.com/trinityrnaseq/trinityrnaseq/

**Comment by abretaud** [comment]
I tried to export PATH in build.sh, but it is reset when build is finished:
```
BUILD START: trinity-2.1.1-2
[.......]
+ cd /home/abretaud/miniconda2/envs/_build/bin
+ ln -s /home/abretaud/miniconda2/envs/_build/opt/trinity-2.1.1/Trinity Trinity
+ export PATH=/home/abretaud/miniconda2/envs/_build/opt/trinity-2.1.1/util/:/home/abretaud/miniconda2/envs/_build/opt/trinity-2.1.1/Analysis/DifferentialExpression/:/home/abretaud/miniconda2/envs/_build/bin:/home/abretaud/.planemo/bin:/home/abretaud/.linuxbrew/bin:/home/abretaud/.linuxbrew/bin:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/home/abretaud/.local/bin:/home/abretaud/bin
+ PATH=/home/abretaud/miniconda2/envs/_build/opt/trinity-2.1.1/util/:/home/abretaud/miniconda2/envs/_build/opt/trinity-2.1.1/Analysis/DifferentialExpression/:/home/abretaud/miniconda2/envs/_build/bin:/home/abretaud/.planemo/bin:/home/abretaud/.linuxbrew/bin:/home/abretaud/.linuxbrew/bin:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/home/abretaud/.local/bin:/home/abretaud/bin
[.......]
BUILD END: trinity-2.1.1-2
TEST START: trinity-2.1.1-2
[.......]
+ echo /home/abretaud/miniconda2/envs/_test:/home/abretaud/miniconda2/envs/_test/bin:/home/abretaud/miniconda2/envs/_test/bin:/home/abretaud/.planemo/bin:/home/abretaud/.linuxbrew/bin:/home/abretaud/.linuxbrew/bin:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/home/abretaud/.local/bin:/home/abretaud/bin
/home/abretaud/miniconda2/envs/_test:/home/abretaud/miniconda2/envs/_test/bin:/home/abretaud/miniconda2/envs/_test/bin:/home/abretaud/.planemo/bin:/home/abretaud/.linuxbrew/bin:/home/abretaud/.linuxbrew/bin:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/home/abretaud/.local/bin:/home/abretaud/bin
+ Trinity --cite
+ align_and_estimate_abundance.pl --help
+ [[ 127 == 255 ]]
TESTS FAILED: trinity-2.1.1-2
```
In meta.yaml, I wrote this to show the PATH content:
```
test:
    commands:
        - "echo $PATH && Trinity --cite &> /dev/null"
```

Here is the error I get when running run_DE_analysis.pl from a symlink
```
$ /tmp/run_DE_analysis.pl -h
Can't locate Fasta_reader.pm in @INC (you may need to install the Fasta_reader module) (@INC contains: /tmp/../../PerlLib /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .) at /tmp/run_DE_analysis.pl line 11.
```

File: packages/package_trinity_2_1_1/tool_dependencies.xml [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+                </action>
+                <action type="move_directory_files">
+                    <source_directory>util</source_directory>
+                    <destination_directory>$INSTALL_DIR/util</destination_directory>
+                </action>
+
+                <action type="shell_command">rm $INSTALL_DIR/trinity-plugins/*tar.gz</action>
+
+                <action type="set_environment">
+                    <environment_variable name="PATH" action="prepend_to">$INSTALL_DIR/</environment_variable>
```
Comment (nsoranzo) [inline packages/package_trinity_2_1_1/tool_dependencies.xml]: I think you need to add also `Analysis/DifferentialExpression` to `PATH` for `run_DE_analysis.pl` and `PtR`.

**Comment by bgruening** [comment]
@abretaud this is tricky, I also tried a few things without success. I would create a PR to bioconda and ask for help.

**Comment by abretaud** [comment]
Here is the PR to bioconda with a somewhat-dirty-but-working solution: [918](https://github.com/bioconda/bioconda-recipes/pull/918)

**Comment by bgruening** [comment]
This is a good solution until the they fix it upstream. Thanks for working on this and sorry to be so nitpicking, I just want to have everything under testing rather sooner than later.

**Comment by abretaud** [comment]
No problem, the new conda system looks very good, it's fine to do it the right way.
For the upstream fix, it was [already done](https://github.com/abretaud/trinityrnaseq/commit/f29615b4b4db30fa8fb7fc6ee1e21e12d7e86a90), so we will just have to remove the find/sed stuff for the next trinity version.

**Comment by bgruening** [comment]
Awesome, this is how Open Source should work!

**Comment by abretaud** [comment]
It looks like the Trinity tests fail because the TRINITY_MEM_OPTIONS env variable is not set
Is there a way to run the tests with a default value?

**Comment by bgruening** [comment]
Let's try some bash magic. `\${GALAXY_SLOTS:-4}` this means that if GALAXY_SLOTS is not set, it will default to 4.

So something like this should work:

 \${TRINITY_MEM_OPTIONS:---max_memory 30G --bflyHeapSpaceMax 30G}

Not sure about the `---`, sorry. I hope this works.

File: packages/package_deseq2_1_10_0/.shed.yml [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+description: Contains a tool dependency definition that downloads and compiles version
+  1.8.2 of DESeq2.
+long_description: |
+  Estimate variance-mean dependence in count data from high-throughput sequencing assays and
+  test for differential expression based on a model using the negative binomial distribution
+
+  http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html
+name: package_deseq2_1_8_2
+owner: iuc
+remote_repository_url: https://github.com/galaxyproject/tools-iuc/tree/master/packages/package_deseq2_1_8_2
```
Comment (nsoranzo) [inline packages/package_deseq2_1_10_0/.shed.yml]: s/1_8_2/1_10_0/

File: packages/package_deseq2_1_10_0/.shed.yml [hunk 17]
```diff
@@ -0,0 +1,13 @@
+categories:
+- Tool Dependency Packages
+description: Contains a tool dependency definition that downloads and compiles version
+  1.8.2 of DESeq2.
+long_description: |
+  Estimate variance-mean dependence in count data from high-throughput sequencing assays and
+  test for differential expression based on a model using the negative binomial distribution
+
+  http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html
+name: package_deseq2_1_8_2
```
Comment (nsoranzo) [inline packages/package_deseq2_1_10_0/.shed.yml]: s/1_8_2/1_10_0/

File: packages/package_deseq2_1_10_0/.shed.yml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+categories:
+- Tool Dependency Packages
+description: Contains a tool dependency definition that downloads and compiles version
+  1.8.2 of DESeq2.
```
Comment (nsoranzo) [inline packages/package_deseq2_1_10_0/.shed.yml]: s/1.8.2/1.10.0/

File: tools/trinity/run_DE_analysis.xml [hunk 19]
```diff
@@ -0,0 +1,145 @@
+<tool id="run_DE_analysis" name="Differential expression analysis" version="2.1.1">
+    <description>using a Trinity assembly</description>
+    <requirements>
+        <requirement type="package" version="2.1.1">trinity</requirement>
+        <!-- odering is crucial, otherwise R will override the ENV variables from deseq2 -->
+        <requirement type="package" version="1.8.2">deseq2</requirement>
```
Comment (nsoranzo) [inline tools/trinity/run_DE_analysis.xml]: I think you want to use "1.10.0" here and in `tools/trinity/tool_dependencies.xml` .

**Comment by nsoranzo** [comment]
> So something like this should work:
>
> \${TRINITY_MEM_OPTIONS:---max_memory 30G --bflyHeapSpaceMax 30G}

Only `--max_memory` is required for Trinity to run.

**Comment by abretaud** [comment]
Right, fixed
The remaining failed tests should pass when [936](https://github.com/bioconda/bioconda-recipes/pull/936) is merged (finger crossed!)

**Comment by nsoranzo** [comment]
I think you still need to fix the deseq2 requirement in `tools/trinity/run_DE_analysis.xml` .

**Comment by abretaud** [comment]
Oh yes, thanks!

**Comment by nsoranzo** [comment]
If I run on my workstation:
```
$ ./Trinity --seqType fa --max_memory 30G --single foo.fa
...
* Running CMD: /opt/trinityrnaseq_trinityrnaseq/trinity-plugins/jellyfish/bin/jellyfish count -t 2 -m 25 -s 4601750669  --canonical  single.fa
terminate called after throwing an instance of 'jellyfish::large_hash::array_base<jellyfish::mer_dna_ns::mer_base_static<unsigned long, 0>, unsigned long, atomic::gcc, jellyfish::large_hash::array<jellyfish::mer_dna_ns::mer_base_static<unsigned long, 0>, unsigned long, atomic::gcc, allocators::mmap> >::ErrorAllocation'
  what():  Failed to allocate 18407002680 bytes of memory
Aborted (core dumped)
Trinity run failed. Must investigate error above.
```

So I guess that this is failing also on the Travis build hosts. I'd suggest to decrease the default to e.g. 1G.

**Comment by nsoranzo** [comment]
It seems `express` may not be in conda, see the Travis logs:
```
galaxy.tools.deps DEBUG 2016-03-04 13:12:28,341 Find dependency express version 1.5.1
galaxy.tools.deps DEBUG 2016-03-04 13:12:28,341 Resolver conda returned <galaxy.tools.deps.resolvers.NullDependency object at 0x7f7298902290> (isnull? True)
galaxy.tools.deps DEBUG 2016-03-04 13:12:28,341 Resolver conda returned <galaxy.tools.deps.resolvers.NullDependency object at 0x7f7298902290> (isnull? True)
galaxy.tools.deps WARNING 2016-03-04 13:12:28,342 Failed to resolve dependency on 'express', ignoring
```
which cause the failure of the third test of the `align_and_estimate_abundance` tool:
```
|  ERROR, cannot find express in PATH setting: /tmp/tmp2kn5qI/job_working_directory/000/27/conda-env/opt/trinity-2.1.1/util/../trinity-plugins/rsem:/tmp/tmp2kn5qI/job_working_directory/000/27/conda-env/bin:/home/travis/virtualenv/python2.7.9/bin:/home/travis/bin:/home/travis/.local/bin:/home/travis/.gimme/versions/go1.4.2.linux.amd64/bin:/home/travis/.rvm/gems/ruby-1.9.3-p551/bin:/home/travis/.rvm/gems/ruby-1.9.3-p551@global/bin:/home/travis/.rvm/rubies/ruby-1.9.3-p551/bin:/opt/python/2.6.9/bin:/opt/python/2.7.9/bin:/opt/python/3.2.5/bin:/opt/python/3.3.5/bin:/opt/python/3.4.2/bin:/opt/python/3.5.0/bin:/opt/python/pypy-2.6.1/bin:/opt/python/pypy3-2.4.0/bin:/usr/local/phantomjs/bin:/home/travis/.nvm/v0.10.36/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/travis/.rvm/bin
|  Please be sure bowtie and express are installed and the utilities samtools bowtie-build bowtie express are available via your PATH setting.
```

**Comment by abretaud** [comment]
Yep, it should be ok when conda [PR](https://github.com/bioconda/bioconda-recipes/pull/936) is merged

**Comment by abretaud** [comment]
Is there a way to relaunch the tests now that the PR has been merged?

**Comment by bgruening** [comment]
Closing and opening the PR should work for you, but I can also trigger them. Give me a sec.

**Comment by bgruening** [comment]
Last 5 tests are failing....

**Comment by nsoranzo** [comment]
Actually it's the last 4 tests, but only the last 2 seems to be really failing, for the other two there are no FAIL sections.

**Comment by abretaud** [comment]
Yep, I'm going to look into it

**Comment by nsoranzo** [comment]
One problem may be the dependency on `limma` 3.25.3 in `run_DE_analysis` tool. In bioconda the package is called `bioconductor-limma` and it is at version 3.26.7, right?

**Comment by bgruening** [comment]
@abretaud if this is too much work for you. I'm ok with merging. It would be wonderful if we can get the tests working but until now it is not requirement by the IUC.

**Comment by abretaud** [comment]
Hum, I see 2 problems to solve:
- the limma/bioconductor-limma package: I'm afraid it will need to be updated both in the toolshed and in bioconda: they are not in the same version on eac one, and the src url is no longer available in bioconda
- some harcoded path in Trinity code, I think a sed hack could fix it in bioconda recipe

I think it is worth making a last(!) effort before merging, hopefully I'll find some time today

**Comment by bgruening** [comment]
Seems that trinity can now be installed but still is not creating any output :(

**Comment by abretaud** [comment]
Yep, still a [dep problem](https://github.com/bioconda/bioconda-recipes/pull/1119)

**Comment by abretaud** [comment]
Hum, I don't understand: looking at the travis log, fastool is missing when running the trinity tests, but it should be present in the "__package__trinity@__version__2.1.1" conda env as it is installed at the beginning...
And on my machine the tests passed.

**Comment by bgruening** [comment]
@abretaud in my `/home/gruening/miniconda2/envs/trinity/bin` is not fastool.

**Comment by abretaud** [comment]
That's strange, I feel like I'm missing something obvious somewhere :/
Here is what I do:
```
[abretaud@lapinou trinity]$ conda create --name trinity_fresh2 python=3 trinity=2.1.1
Fetching package metadata: ........
Solving package specifications: ...........

Package plan for installation in environment /home/abretaud/miniconda2/envs/trinity_fresh2:

The following NEW packages will be INSTALLED:

    collectl:      4.0.4-0      
    fastool:       0.1.4-0      
    htslib:        1.3-0        
    java-jdk:      8.0.45-0     
    jellyfish:     2.2.3-0      
    libgcc:        4.8.5-1      
    openssl:       1.0.2g-0     
    parafly:       r2013_01_21-0
    perl-threaded: 5.22.0-9     
    pip:           8.1.1-py35_0 
    python:        3.5.1-0      
    readline:      6.2-2        
    samtools:      1.3-0        
    setuptools:    20.3-py35_0  
    slclust:       02022010-1   
    sqlite:        3.9.2-0      
    tk:            8.5.18-0     
    trimmomatic:   0.36-3       
    trinity:       2.1.1-6      
    wheel:         0.29.0-py35_0
    xz:            5.0.5-1      
    zlib:          1.2.8-0      

Proceed ([y]/n)? 

Extracting packages ...
[      COMPLETE      ]|##################################################################################################################################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|##################################################################################################################################################################################################################################################| 100%
#
# To activate this environment, use:
# $ source activate trinity_fresh2
#
# To deactivate this environment, use:
# $ source deactivate
#
[abretaud@lapinou trinity]$ ll /home/abretaud/miniconda2/envs/trinity_fresh2/bin/fastool 
-rwxr-xr-x 3 abretaud xxxx 17920 10 mars  14:28 /home/abretaud/miniconda2/envs/trinity_fresh2/bin/fastool
```

**Comment by bgruening** [comment]
@abretaud party time!
I fixed fastool in bioconda and now everything works! If you are ready to merge this, I'm too :)

Thanks for your patience and helping me with the conda stuff. This was a really nice co-work! Thank you so much!

**Comment by abretaud** [comment]
Wouhou! Excellent!
wouhou!

I'm ready to merge yes! Thank you for all your help!
(and now, let's move on to stacks)

**Comment by bgruening** [comment]
Do you have a twitter account? Let's go to stacks! :)

**Comment by abretaud** [comment]
Yep, I am a modern man, I'm on twitter: @abretau

**Comment by abretaud** [comment]
Hum, it looks like it's not yet uploaded to the main toolshed: is there something I need to do before, or is it just a normal delay?

**Comment by bgruening** [comment]
Normal delay ... I'm working on an automatic TS upload if all tests are passing, but this is not activated yet. Still a manual process.

**Comment by abretaud** [comment]
Ok, no problem, thanks!

**Comment by bgruening** [comment]
TTS is updated.

**Comment by abretaud** [comment]
Would it be possible to push this to MTS? I will need it in a workflow in the coming days

**Comment by bgruening** [comment]
Done!

**Comment by abretaud** [comment]
Excellent! Thanks
